packages/nextjs/app/Home/builders/page.tsx

```diff
@@ -0,0 +1,126 @@
+"use client";
+
+import { useEffect, useState } from "react";
+import { useScaffoldEventHistory, useScaffoldReadContract } from "~~/hooks/scaffold-eth";
+
+const styles = `
+    .glass-effect {
+        background: rgba(255, 255, 255, 0.05);
+        border-radius: 16px;
+        box-shadow: 0 4px 30px rgba(0, 0, 0, 0.1);
+        backdrop-filter: blur(7px);
+        -webkit-backdrop-filter: blur(7px);
+        border: 1px solid rgba(255, 255, 255, 0.1);
+    }
+`;
+
+export default function Home() {
+  const { data: totalCounter } = useScaffoldReadContract({
+    contractName: "BatchRegistry",
+    functionName: "checkedInCounter",
+  });
+
+  const parsedCounter = typeof totalCounter === "bigint" ? totalCounter.toString() : totalCounter;
+
+  return (
+    <>
+      <style>{styles}</style>
+      <div className="min-h-screen bg-gradient-to-br from-gray-900 to-gray-800 text-white p-8">
+        <div className="glass-effect rounded-3xl p-8 mb-6">
+          <h1 className="text-4xl font-extrabold text-center mb-6 text-transparent bg-clip-text bg-gradient-to-r from-blue-400 to-purple-500">
+            Checked-In Builders: <span className="text-white">{parsedCounter ?? "Loading..."}</span>
+          </h1>
+          <p className="text-center text-gray-300 italic">Discover the builders who&apos;ve joined the network.</p>
+        </div>
+        <Members />
+      </div>
+    </>
+  );
+}
+
+interface Event {
+  args?: {
+    builder?: string;
+    checkInContract?: string;
+    first?: boolean;
+  };
+}
+
+function Members() {
+  const {
+    data: events,
+    isLoading,
+    error,
+  } = useScaffoldEventHistory({
+    contractName: "BatchRegistry",
+    eventName: "CheckedIn",
+    fromBlock: 131801148n,
+  });
+
+  const [builders, setBuilders] = useState<{ builder: string; checkInContract: string }[]>([]);
+
+  useEffect(() => {
+    console.log("Raw events data:", events);
+
+    if (!events || !Array.isArray(events)) {
+      console.error("Expected events to be an array, but got:", typeof events, events);
+      return;
+    }
+
+    const extractedBuilders = events
+      .filter((event: Event) => event.args?.first === true)
+      .map((event: Event) => ({
+        builder: event.args!.builder as string,
+        checkInContract: event.args!.checkInContract as string,
+      }));
+
+    setBuilders(extractedBuilders);
+    console.log("Filtered builders:", extractedBuilders);
+  }, [events]);
```

Comment:
There's a very common pitfall while developing reactive applications: creating more states than you need.  

You already have a state for the events, so you don't need a new state to save the builders. The builders are inside the events, just in a different format. Every time you change the events, the builders should change, and no developer should be able to manually change the builders, which is currently possible in your code.  

So instead of using `useState` and `useEffect`, use `useMemo` to transform events into builders in a new variable.  
Try to always remember: if it's possible to build something without `useState` and `useEffect`, it will be easier to maintain.  

Take a look at the `useMemo` documentation and ask me if you have any doubts.